### ISSUE TITLE
conda SSL errors during `bioconda-utils build`

### ISSUE THREAD

**Issue by unode** [bioconda/bioconda-recipes#9225]
Hi everyone,

I'm trying to package and test a new addition.
Using the `bootstrap.py` method and running linting reports no errors.
However during build I'm running into SSL errors that are causing failures on several stages of the build process.

Lately I've also seen a lot of SSL failures in conda and general recommendation to set `ssl_verify = false`.

Regardless, are there any workarounds?

Some logs below:
```
16:18:06 BIOCONDA ERROR COMMAND FAILED: docker run --net host --rm -v /tmp/tmpslpy1vxt/build_script.bash:/opt/build_script.bash -v /var/tmp/miniconda/miniconda/conda-bld:/opt/host-conda-bld -v /home/ralves/projects/bioconda-recipes/recipes/motus2:/opt/recipe -e CONDA_DEFAULT_ENV=root -e LANG=en_GB.UTF-8 -e LC_ALL=en_GB.UTF-8 -e CONDA_ROOT=/var/tmp/miniconda/miniconda -e CONDA_ALLOW_SOFTLINKS=false -e CONDA_PY=27 -e CONDA_HTSLIB=1.7 -e CONDA_BOOST=1.64 -e CONDA_R=3.4.1 -e CONDA_PERL=5.22.0 -e CONDA_NPY=112 -e CONDA_NC
URSES=5.9 -e CONDA_GSL=2.2 -e CONDA_GMP=5.1 -e CONDA_HDF5=1.8.17 -e MACOSX_DEPLOYMENT_TARGET=10.9 -e CONDA_ZLIB=1.2.11 -e CONDA_BZIP2=1.0 -e CONDA_XZ=5.2 -e CONDA_BAMTOOLS=2.4.1 -e HOST_USER_ID=1000 tmp-bioconda-builder /bin/bash /opt/build_script.bash
16:18:06 BIOCONDA ERROR STDOUT+STDERR:                                                                                                                                                                                                                                           
bash: cannot set terminal process group (-1): Inappropriate ioctl for device                 
bash: no job control in this shell                                                                                                                                                                                                                                               
bash: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8)                                                                                                                                                                                                             
/bin/bash: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8)                                                                                                              
/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8)                                                                                                                                                                                                          
/bin/sh: warning: setlocale: LC_ALL: cannot change locale (en_GB.UTF-8)                                                                                                                                                                                                          
INFO:filelock:Lock 140028253261608 acquired on /opt/conda/locks/L29wdC9jb25kYS9wa2dz                                                                                                                                                                                             
INFO:filelock:Lock 140028253260600 acquired on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                                             
INFO:filelock:Lock 140028253261608 released on /opt/conda/locks/L29wdC9jb25kYS9wa2dz                                                                                                                                                                                             
INFO:filelock:Lock 140028253260600 released on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                        
INFO:filelock:Lock 140028253260600 acquired on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                                             
INFO:filelock:Lock 140028253260600 acquired on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                                             
INFO:filelock:Lock 140028253260600 released on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                        
INFO:filelock:Lock 140028253260656 acquired on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                                             
INFO:filelock:Lock 140028253260656 acquired on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                                             
INFO:filelock:Lock 140028253260656 released on /opt/conda/locks/L29wdC9jb25kYS9jb25k                                                                                                                                                                                             
+ source /opt/conda/bin/activate /opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p
lacehold_placeh                                                                                                                                                                                                                                                                  
+ python setup.py                                                                                                                                                                                                                                           
 ------------------------------------------------------------------------------                                                                                                                                                                                                  
|                              SETUP MOTUS TOOL                                |                                                                                                                                                                                                 
 ------------------------------------------------------------------------------
Download the compressed motus database (~1Gb)                                                                 
Traceback (most recent call last):                                                                                                                                                                                                                                               
  File "setup.py", line 190, in <module>                                                                                                                                                                                                                                         
    status = main()                                                                                
  File "setup.py", line 89, in main                                                      
    u = urllib2.urlopen(link_db)                                                                                                                                                                                                                                                 
  File "/opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pytho
n2.7/urllib2.py", line 154, in urlopen                                                            
    return opener.open(url, data, timeout)                                                                                                                                                                                                                                       
  File "/opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pytho
n2.7/urllib2.py", line 429, in open                                                 
    response = self._open(req, data)                                                                                                                                                                                                                                             
  File "/opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pytho
n2.7/urllib2.py", line 447, in _open                                                                                 
    '_open', req)                                                                                                                                                                                                                                                                
  File "/opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pytho
n2.7/urllib2.py", line 407, in _call_chain                                                                                                                                                                                                                                       
    result = func(*args)                                                                                                                                                                                                                                                         
  File "/opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pytho
n2.7/urllib2.py", line 1241, in https_open                                                                                                                                                                                                                                       
    context=self._context)                                                                                                                                                                                                                                                         File "/opt/conda/conda-bld/motus2_1528467431163/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/python2.7/urllib2.py", line 1198, in do_open                                                                                                                                                                                                                                         
    raise URLError(err)                                                                                                                                                                                                                                                          
urllib2.URLError: <urlopen error [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:726)>                                                                                                                                                                        
                                                                                                                                                                                                                                                                                 WARNING: conda-build appears to be out of date. You have version 2.1.18 but the                                                                                                                                                                                                  
latest version is 3.10.7. Run                                                                                                                                                                                                                                                                                                                                     
conda update -n root conda-build                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  
to get the latest version.
```
and later:
```
BUILD START: motus2-2.0.0-py27_0
updating index in: /opt/conda/conda-bld/linux-64
updating index in: /opt/conda/conda-bld/noarch
    
The following NEW packages will be INSTALLED:
    
    ca-certificates: 2018.4.16-0       conda-forge
    certifi:         2018.4.16-py27_0  conda-forge
    libgcc-ng:       7.2.0-hdf63c60_3  defaults
    ncurses:         5.9-10            conda-forge
    openssl:         1.0.2o-0          conda-forge
    pip:             9.0.3-py27_0      conda-forge
    python:          2.7.15-0          conda-forge
    readline:        7.0-0             conda-forge
    setuptools:      39.2.0-py27_0     conda-forge
    sqlite:          3.20.1-2          conda-forge
    tk:              8.6.7-0           conda-forge
    wheel:           0.31.0-py27_0     conda-forge
    zlib:            1.2.11-h470a237_3 conda-forge

Source cache directory is: /opt/conda/conda-bld/src_cache
Downloading source to cache: 2.0.0-rc1.tar.gz
Downloading https://github.com/motu-tool/mOTUs_v2/archive/2.0.0-rc1.tar.gz
Success
Extracting download
Package: motus2-2.0.0-py27_0
source tree in: /opt/conda/conda-bld/motus2_1528467431163/work/mOTUs_v2-2.0.0-rc1
Traceback (most recent call last):
  File "/opt/conda/bin/conda-build", line 6, in <module>
    sys.exit(conda_build.cli.main_build.main())
  File "/opt/conda/lib/python3.6/site-packages/conda_build/cli/main_build.py", line 342, in main
    execute(sys.argv[1:])
  File "/opt/conda/lib/python3.6/site-packages/conda_build/cli/main_build.py", line 333, in execute
    noverify=args.no_verify)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/api.py", line 97, in build
    need_source_download=need_source_download, config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1524, in build_tree
    config=config)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/build.py", line 1147, in build
    utils.check_call_env(cmd, env=env, cwd=src_dir)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/utils.py", line 628, in check_call_env
    return _func_defaulting_env_to_os_environ(subprocess.check_call, *popenargs, **kwargs)
  File "/opt/conda/lib/python3.6/site-packages/conda_build/utils.py", line 624, in _func_defaulting_env_to_os_environ
    return func(_args, **kwargs)
  File "/opt/conda/lib/python3.6/subprocess.py", line 291, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/bin/bash', '-x', '-e', '/opt/conda/conda-bld/motus2_1528467431163/work/mOTUs_v2-2.0.0-rc1/conda_build.sh']' returned non-zero exit status 1.
```

and also:

```
...
+ source /opt/conda/bin/activate /opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_p
lacehold_placeh                                                                                                                                                                                                                                                                  
+ python setup.py
 ------------------------------------------------------------------------------
|                              SETUP MOTUS TOOL                                |
 ------------------------------------------------------------------------------
Download the compressed motus database (~1Gb)
Traceback (most recent call last):
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/urllib/request.py", line 1254, in do_open
    h.request(req.get_method(), req.selector, req.data, headers)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/http/client.py", line 1107, in request
    self._send_request(method, url, body, headers)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/http/client.py", line 1152, in _send_request
    self.endheaders(body)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/http/client.py", line 1103, in endheaders
    self._send_output(message_body)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/http/client.py", line 934, in _send_output
    self.send(msg)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/http/client.py", line 877, in send
    self.connect()
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/http/client.py", line 1261, in connect
    server_hostname=server_hostname)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/ssl.py", line 385, in wrap_socket
    _context=self)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/ssl.py", line 760, in __init__
    self.do_handshake()
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/ssl.py", line 996, in do_handshake
    self._sslobj.do_handshake()
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/ssl.py", line 641, in do_handshake
    self._sslobj.do_handshake()
ssl.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed (_ssl.c:719)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "setup.py", line 190, in <module>
    status = main()
  File "setup.py", line 111, in main
    save_f(link_db, db_name)
  File "setup.py", line 58, in save_f
    urllib.request.urlretrieve(url, filename, reporthook)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/urllib/request.py", line 188, in urlretrieve
    with contextlib.closing(urlopen(url, data)) as fp:
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/urllib/request.py", line 163, in urlopen
    return opener.open(url, data, timeout)
  File "/opt/conda/conda-bld/motus2_1528467489852/_b_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placeh/lib/pyth$
n3.5/urllib/request.py", line 466, in open
    response = self._open(req, data)
... (truncated)
```

**Comment by daler** [comment]
@unode it looks like the tool is downloading a database from within its `setup.py`. We probably don't want to do this since it will make the package large and consume storage in the channel. Typically for these sorts of situations we use post-link scripts (like the bioconductor data packages) to install required databases separately.

Would you please submit a PR for this recipe so we can take a look?

**Comment by unode** [comment]
@daler will do. Haven't yet since what I had wasn't final.
I'm still tweaking the conda recipe and trying to get it to build.

About the large data blob, yes this is the case. I wanted to open another issue on that point (blob is ~1GB gzipped). Didn't find any guidance on how to bundle data in https://bioconda.github.io/contribute-a-recipe.html so I wasn't sure how to proceed.

You mentioned bioconductor packages. Could you point out a good-practice example?
I'll proceed with introducing this change and do the pull request after.

**Comment by daler** [comment]
@unode thanks. Here's an example using post-link:

https://github.com/bioconda/bioconda-recipes/tree/master/recipes/bioconductor-txdb.dmelanogaster.ucsc.dm6.ensgene

Here's an example that strips out the data from the tarball before installing, and then provides an additional script for downloading the example data:

https://github.com/bioconda/bioconda-recipes/tree/master/recipes/spectacle

**Comment by unode** [comment]
@daler ping

In case you missed the references above.

**Comment by unode** [comment]
Fixed in #9239
